diff --git a/vendor/magento/module-import-export/Model/Import/Entity/AbstractEav.php b/vendor/magento/module-import-export/Model/Import/Entity/AbstractEav.php
index 40124817..80c63419 100644
--- a/vendor/magento/module-import-export/Model/Import/Entity/AbstractEav.php
+++ b/vendor/magento/module-import-export/Model/Import/Entity/AbstractEav.php
@@ -224,12 +224,7 @@ abstract class AbstractEav extends \Magento\ImportExport\Model\Import\AbstractEn
                     foreach ($value as $innerOption) {
                         // skip ' -- Please Select -- ' option
                         if (strlen($innerOption['value'])) {
-                            if ($attribute->isStatic()) {
-                                $options[strtolower($innerOption[$index])] = $innerOption['value'];
-                            } else {
-                                // Non-static attributes flip keys an values
-                                $options[$innerOption['value']] = $innerOption[$index];
-                            }
+                            $options[mb_strtolower($innerOption[$index])] = $innerOption['value'];
                         }
                     }
                 }